guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+
+```rust
+trait Service<R> {
+    async fn ready(&mut self);
+}
+```
+
+`ready` would be an async function that completes when the service has capacity
+enough to receive one new request. We would then require users to first call
+`service.ready().await` before doing `service.call(request).await`.
+
+Separating "calling the service" from "reserving capacity" also unlocks new use
+cases such as being able to maintain a set of "ready services" that we keep up
+to date in the background such that when a request arrives we already have a
+ready service to send it to and don't have to first wait for it to become ready.
+
+With this setup `ConcurrencyLimit` could track capacity inside `ready` and not
+allow users to call `call` until there is sufficient capacity.
+
+`Service`s that don't care about capacity can just return immediately from
+`ready` or delegate to some inner `Service` they're wrapping.
+
+However we still cannot define async functions inside traits. We could add
+another associated type to `Service` called `ReadyFuture` but having to return a
+`Future` will give us the same lifetime issues we've run into before. It would
+be nice if there was someway around that.
+
+Instead we can take some inspiration from the `Future` trait and define a method
+called `poll_ready`:
+
+```rust
+use std::task::{Context, Poll};
+
+trait Service<R> {
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<()>;
+}
+```
+
+This means if the service is at capacity `poll_ready` will return
+`Poll::Pending` and somehow ensure to be woken up using `Context` when capacity
+becomes available. At that point `poll_ready` can be called again and if it
+returns `Poll::Ready(())` then capacity would be reserved and `call` can be
+called.
+
+It would still be possible get a `Future` that resolves when capacity is
+available using something like [`futures::future::poll_fn`] (or
+[`tower::ServiceExt::ready`]).
+
+This concept of services communicating with their callers about their capacity
+is called "backpressure propagation". The fundamental idea is that you shouldn't
+send a request to a service that doesn't have the capacity to handle it. Instead
+you should wait, call another service, or somehow scale up the number of
+services available. You can learn more about the general concept of backpressure
+[here][backpressure].
+
+Finally, it might also be possible for some error to happen while reserving
+capacity so `poll_ready` probably should return `Poll<Result<(), Self::Error>>`.
+With this change we've now arrived at the complete `tower::Service` trait:
+
+```rust
+pub trait Service<Request> {
+    type Response;
+    type Error;
+    type Future: Future<Output = Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<(), Self::Error>>;
+
+    fn call(&mut self, req: Request) -> Self::Future;
+}
+```
+
+Middlewares that care about backpressure is pretty rare but it does enable some
```

Comment:
buffering is the other big one! This is what like TCP does with its kernel buffer.

Reply:
I'm not familiar with that. Would you mention that, and if so how would you phrase it?